### DISCUSSION TITLE
How to make 'Planner' Style Calendar View (Week w/ Hourly breakdown)

### DISCUSSION THREAD

**Discussion by zimmra** [Budibase/budibase#5551] [Help] [answered]
Before I really buckle down and try to figure this out, does anyone know of an easy way to make a weekly calendar view?
Looking for an hourly breakdown on each day of reported time entries, similar to image below. 

For each block you see in the image I have
Start Date/Time
End Date/Time
Duration of entry in hours

![image](https://user-images.githubusercontent.com/28514085/164836418-fe986215-6ffa-4783-9257-af648e038083.png)

The 'quick and dirty' way I would probably employ is using cards, but would like to actually plot it on a chart/calendar if possible

**Comment by melohagan** [comment, marked as answer]
Hey @zimmra 

Whilst there is no handy weekly calendar component, as you suggested it can be built using cards. It is fairly complex, but hopefully this guide will be useful. 
 
![timesheet](https://user-images.githubusercontent.com/101575380/165951243-f1d2a311-c52f-45dd-a29f-6f35393845d9.gif)

### Create a new app
To make use of existing functionality, I created a new app using the *Employee Timesheet Template*
https://budibase.com/professional-services/templates/employee-timesheet-template/

### Data
We will add two tables which will be used for the sole purpose of formatting our hourly rows and daily cards. This will allow us to use repeaters, which will greatly reduce the amount of duplication, both for styles and code. 

> One Container Row -> Many Day Cards

![Screenshot 2022-04-29 at 14 34 02](https://user-images.githubusercontent.com/101575380/165954695-b23d8a5b-1b72-4a2f-8b9d-20e7aa9d4c3e.png)

![Screenshot 2022-04-29 at 14 34 44](https://user-images.githubusercontent.com/101575380/165954788-28cf0ac2-47df-48fa-9f1b-eddd3250d041.png)

### Design - Form Inputs
On the **/** home screen, add a form with a row-direction <a href="https://docs.budibase.com/docs/container">container</a>, as well as inputs for the calendar.
![Screenshot 2022-04-29 at 14 39 32](https://user-images.githubusercontent.com/101575380/165955648-5688aa59-d21e-45d0-aaf7-8860ac0de26c.png)

**Define actions** for the 'Apply' button, where we will **Update State** and set *TimesheetDate* to an array of our fields. 
```js
return [
	$("Form.Fields.StartDate"),
	$("Form.Fields.Month"),
	$("Form.Fields.Year")
]
```
An array is used here so that the state, and screen, is only updated once and not three times.

### Design - Column Headers
Below the form (not nested within), add a row direction container, and add one Card for the week title, and five Stat Cards for the day titles.  
![Screenshot 2022-04-29 at 14 45 58](https://user-images.githubusercontent.com/101575380/165957017-148be527-82ff-4340-bccd-de0f23be1975.png)

As a little custom CSS was used, I recommend creating the first day card, and then duplicating it four more times after the custom style is added, which is simply:
```css
min-width: 17%;
border: none;
```
The border is entirely optional, but the *min-width* is chosen to be a sixth of the container, as there are six elements including the week card. 
Enter the title for each day of the week.

For the week title, add this snippet of JavaScript:
```js
let timesheetDate = $("State.TimesheetDate")
if (!timesheetDate) return "-"

let startDate = timesheetDate[0]
let month = timesheetDate[1]
let year = timesheetDate[2]
let date = new Date(year, month-1, startDate)
const day = date.getDay();
const weekOfMonth = Math.ceil((startDate - 1 - day) / 7)+1;
return "Week " + weekOfMonth

```
This simply determines the week number for the input start date, month and year.

### Design - Calendar Grid
This is where the formatting tables from the data step will be leveraged.

Underneath the Header Container (not nested within), add a Data Provider for the *Container Row Format* table.
Sort by the hour ascending to make sure the rows go from 6AM through to 6PM:
![Screenshot 2022-04-29 at 14 57 12](https://user-images.githubusercontent.com/101575380/165958893-9c4761dd-cec7-4a57-b513-75f6cff7615b.png)

Nest a repeater and nest again a row directional container. This will represent an hourly row. 
The end tree should look like:
![Screenshot 2022-04-29 at 14 59 21](https://user-images.githubusercontent.com/101575380/165959260-a9e187dc-268d-468d-b5f8-b1a5e5a8b0ac.png)

Within the container, add a Card that will represent the title of the hour, e.g. 6AM or 5PM. 
To determine AM or PM, a little JavaScript is used in the title binding:
```js
let hour = $("Repeater.Container Row Format.Hour");
return hour + (hour < 12 ? "AM" : "PM")
```

And add a little custom CSS:
```css
min-width: 14%;
min-height: 100px;
```

Below the time card, add another row directional container that will be used to hold the day cards for each hour. 
Note this container will need *Width* to be set to 200%. I'm not quite sure the reason for this, possibly because it's nested within another container.
![Screenshot 2022-04-29 at 15 04 25](https://user-images.githubusercontent.com/101575380/165960115-194d61a6-e5b5-4663-a65c-6ac5b95e3fb6.png)

Inside the day card container, add a Data Provider. This time the source will be the days relationship: 
![Screenshot 2022-04-29 at 15 05 46](https://user-images.githubusercontent.com/101575380/165960371-277f8998-d071-482f-84b6-87707915158b.png)
More on related data sources here: https://docs.budibase.com/docs/fetching-data#displaying-relational-data

Also **Edit custom CSS** and make sure the width is full `width: 100%;`

As with the other Data Provider, we want to sort by the *Day Enum* ascending to make sure the day data is ordered Monday through Friday.
![Screenshot 2022-04-29 at 15 10 53](https://user-images.githubusercontent.com/101575380/165961301-7101f8aa-d90a-47cf-a3d3-a698b7a31964.png)

Add a repeater for the day provider, so that we can loop over each day of the week.

Finally add in two cards: a blank stat card for when that time slot is not assigned, and a coloured detail stat card for when that time slot is filled.

For both cards add a little custom CSS:
```css
min-width: 14%;
min-height: 100px;
```
And an extra colour line for the detail stat card: `background-color: aqua;`

### Design - Functionality
For the detail stat card, add the following JavaScript binding to the title:

**Detail Title Code**
```js
const dayEnum = $("Day Repeater.Day Card Format.Day Enum")
const hourEnum = $("Repeater.Container Row Format.Hour")
const timesheetDate = $("State.TimesheetDate")

if (!timesheetDate) return ""

let month = timesheetDate[1]-1
let year = timesheetDate[2]-1900

if (month && year) {
	const rows = $("Data Provider.Rows")
	.map(row => ({...row, adjustedDate: new Date(row.Date)}))
	.filter(row => row.adjustedDate.getDay() === dayEnum &&
				 row.adjustedDate.getDate() >= timesheetDate[0] &&
				 row.adjustedDate.getDate() < timesheetDate[0]+7 &&
				 row.adjustedDate.getHours() === hourEnum &&
				 row.adjustedDate.getMonth() === month &&
				 row.adjustedDate.getYear() === year)
		
		if (rows?.length > 0) {
			let minutes = "" + rows?.[0]?.adjustedDate.getMinutes()
			minutes = minutes.length === 1 ? "0" + minutes : minutes
			return rows?.[0]?.adjustedDate.getHours() + ":" + minutes
		}
}
return ""
```
To summarise this script, whenever the date of a *Time Log* matches the user input, and matches the hour of the current row, then display the time in digital format, e.g. `06:05`

We also want to show the *Client* that the slot was booked for, and so for the Label JavaScript binding add in this snippet:

**Detail Label Code**
```js
const dayEnum = $("Day Repeater.Day Card Format.Day Enum")
const hourEnum = $("Repeater.Container Row Format.Hour")
const timesheetDate = $("State.TimesheetDate")

if (!timesheetDate) return ""

let month = timesheetDate[1]-1
let year = timesheetDate[2]-1900

if (month && year) {
	const rows = $("Data Provider.Rows")
	.map(row => ({...row, adjustedDate: new Date(row.Date)}))
	.filter(row => row.adjustedDate.getDay() === dayEnum &&
				 row.adjustedDate.getDate() >= timesheetDate[0] &&
				 row.adjustedDate.getDate() < timesheetDate[0]+7 &&
				 hourEnum >= row.adjustedDate.getHours() &&
				 hourEnum <= row.adjustedDate.getHours()+row.Hours &&
				 row.adjustedDate.getMonth() === month &&
				 row.adjustedDate.getYear() === year)
		
		if (rows?.length > 0) {
			let client = rows?.[0]?.Client?.[0].primaryDisplay
			if (client?.length > 15) {
				client = client.substring(0, 15) + "..."
			}
			return client ?? ""
		}
}
return ""
```
The code here is very similar, the difference is that the **primaryDisplay** of the *Time Log* -> *Client* relationship is returned. 
In addition, 
```js
hourEnum >= row.adjustedDate.getHours() &&
hourEnum <= row.adjustedDate.getHours()+row.Hours &&
```
this part gets checks the range of time between the *Time Log* start date and the duration, so that multiple blocks can be filled out. 

The last thing to do is add the **Show** conditions to both cards. The code will be the same as above, but instead of returning a string, return **true**. For the empty string branches, return **false**.

![Screenshot 2022-04-29 at 15 38 32](https://user-images.githubusercontent.com/101575380/165966708-6bed8494-3688-4f42-bef7-2ae97ddea46f.png)
 
The Blank stat card will have the same condition, but the opposite boolean: False.  

App Export: [Timesheet Template-export-1651238226963.txt](https://github.com/Budibase/budibase/files/8591598/Timesheet.Template-export-1651238226963.txt)

### Limitations
- Assumes one client per time block.
- Doesn't support overlap of time blocks.
- A lot of JavaScript is injected, so the performance could be an issue. May need to add a filter to the top level data provider based on the date input to reduce the rows being looped through in each card.
